homeassistant/components/tado/climate.py

```diff
@@ -338,25 +343,63 @@ def __init__(
         self._async_update_home_data()
         self._async_update_zone_data()
 
-    async def async_added_to_hass(self) -> None:
-        """Register for sensor updates."""
-        self.async_on_remove(
-            async_dispatcher_connect(
-                self.hass,
-                SIGNAL_TADO_UPDATE_RECEIVED.format(self._tado.home_id, "home", "data"),
-                self._async_update_home_callback,
-            )
-        )
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Handle updated data from the coordinator."""
+        self._async_update_home_data()
+        self._async_update_zone_data()
+        super()._handle_coordinator_update()
+
+    @callback
+    def _async_update_zone_data(self) -> None:
+        """Load tado data into zone."""
+        self._tado_zone_data = self._tado.data["zone"][self.zone_id]
+
+        # Assign offset values to mapped attributes
+        for offset_key, attr in TADO_TO_HA_OFFSET_MAP.items():
+            if (
+                self._device_id in self._tado.data["device"]
+                and offset_key
+                in self._tado.data["device"][self._device_id][TEMP_OFFSET]
+            ):
+                self._tado_zone_temp_offset[attr] = self._tado.data["device"][
+                    self._device_id
+                ][TEMP_OFFSET][offset_key]
+
+        self._current_tado_hvac_mode = self._tado_zone_data.current_hvac_mode
+        self._current_tado_hvac_action = self._tado_zone_data.current_hvac_action
 
-        self.async_on_remove(
-            async_dispatcher_connect(
-                self.hass,
-                SIGNAL_TADO_UPDATE_RECEIVED.format(
-                    self._tado.home_id, "zone", self.zone_id
-                ),
-                self._async_update_zone_callback,
+        if self._is_valid_setting_for_hvac_mode(TADO_FANLEVEL_SETTING):
+            self._current_tado_fan_level = self._tado_zone_data.current_fan_level
+        if self._is_valid_setting_for_hvac_mode(TADO_FANSPEED_SETTING):
+            self._current_tado_fan_speed = self._tado_zone_data.current_fan_speed
+        if self._is_valid_setting_for_hvac_mode(TADO_SWING_SETTING):
+            self._current_tado_swing_mode = self._tado_zone_data.current_swing_mode
+        if self._is_valid_setting_for_hvac_mode(TADO_VERTICAL_SWING_SETTING):
+            self._current_tado_vertical_swing = (
+                self._tado_zone_data.current_vertical_swing_mode
+            )
+        if self._is_valid_setting_for_hvac_mode(TADO_HORIZONTAL_SWING_SETTING):
+            self._current_tado_horizontal_swing = (
+                self._tado_zone_data.current_horizontal_swing_mode
             )
-        )
+
+    @callback
+    def _async_update_zone_callback(self) -> None:
+        """Load tado data and update state."""
+        self._async_update_zone_data()
+        super().async_write_ha_state()
```

Comment:
So why do we do this write here? This would already be handled by `_handle_coordinator_update` right?